### PR TITLE
Add meeting scheduling link

### LINKED ISSUE #832: Schedule a call link
Most, if not all, of successful placements include a live call of some sorts.

But the conversation UI is pretty terrible at trying to coordinate a time that works for both parties.

image

I'd love a one-click "Schedule a meeting" button for businesses to help streamline this process.

And no, I don't want to rebuilt all of SavvyCal. As an MVP this can be as simple as linking directly to the developer's SavvyCal or Calendly link.

A PR addressing this issue should:

- [ ] Add a new field to developer profiles - scheduling link
- [ ] Show a "Schedule a meeting" button to businesses when viewing a conversation
- [ ] Open the scheduling link in a new tab

### PR DESCRIPTION
**Pull request by jkostolansky**
This PR adds:
- `scheduling_link` to the database table `developers`.
- Scheduling section with link field to developer profile form.
- "Schedule a meeting" link to conversations for businesses.

I haven't added any validation for the `scheduling_link` attribute (yet?). In the issue description there's mentioned support for SavvyCal, Calendly, and Cal.com. But I'm thinking... is it necessary to restrict it? For example, if any of these services support custom domain (and it seems Cal.com supports it on the higher plan), that would be invalid.

I also haven't added basic URL validation for this attribute. I've noticed the current `website` attribute doesn't have any validation, too. And I've also noticed that the fields `website` and `mastodon` don't remove the protocol prefix (`http/s`), so it's somewhat duplicated in the form. I think these issues should be fixed together to maintain consistency.

And I also haven't added other translations (es, pt, fr, zh). I don't speak these languages, but I can use Google Translator if you want.

Screenshots:

CleanShot 2023-03-24 at 12 54 09

CleanShot 2023-03-24 at 12 54 38

Closes #832

### CODE REVIEW

File: app/components/conversations/read_indicator_component.html.erb [hunk 1]
```diff
@@ -1,14 +1,10 @@
...
-<% else %>
-  <div class="flex justify-end my-3 mr-4">
-    <div class="flex items-center justify-center space-x-2 text-sm">
+    <% else %>
       <p class="text-gray-500"><%= t(".unread") %></p>
-    </div>
+    <% end %>
   </div>
-<% end %>
+</div>
```
Comment (jkostolansky) [inline app/components/conversations/read_indicator_component.html.erb:10]: No functional change here, I've just removed some duplicated code by putting the `if` condition deeper.

File: app/assets/images/icons/outline/calendar_days.svg
Comment (joemasilotti) [inline app/assets/images/icons/outline/calendar_days.svg:1]: Can you pease use the [v1 version](https://v1.heroicons.com) of this icon? RailsDevs hasn't upgraded to v2 yet.

File: app/components/conversations/read_indicator_component.html.erb (hunk 1 above)
Comment (joemasilotti) [reply app/components/conversations/read_indicator_component.html.erb:10]: Nice, thank you!

File: app/components/businesses/scheduling_link_component.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  class SchedulingLinkComponent < ApplicationComponent
+    private attr_reader :user, :conversation
+
+    def initialize(user, conversation)
+      @user = user
+      @conversation = conversation
+    end
+
+    def render?
+      user.business && scheduling_link.present?
```
Comment (joemasilotti) [inline app/components/businesses/scheduling_link_component.rb:11]: This should check if the user's business is the same as the business involved in the conversation. I think we have a method on `Conversation` to check that.

**Review by joemasilotti — CHANGES_REQUESTED** [review]
This is looking great so far! Thanks for the quick turnaround.

I'll tweak the copy of the developer form but this is a great start. Thanks for kicking that off.

Can you please add the NEW functionality to this field? Pull the string from `Developers::NewFieldComponent` to get started.

I'd like the design to be a button group, similar to the admin section on a developer profile. Let me know if you'd like to apply that change or if you want me to take the first pass.

Also, I'd love for the Core Developer to be seeded with a scheduling link.

As for your specific questions...

> I haven't added any validation for the scheduling_link attribute (yet?). In the issue description there's mentioned support for SavvyCal, Calendly, and Cal.com. But I'm thinking... is it necessary to restrict it? For example, if any of these services support custom domain (and it seems Cal.com supports it on the higher plan), that would be invalid.

I agree - let's not validate this for now.

> I also haven't added basic URL validation for this attribute. I've noticed the current website attribute doesn't have any validation, too. And I've also noticed that the fields website and mastodon don't remove the protocol prefix (http/s), so it's somewhat duplicated in the form. I think these issues should be fixed together to maintain consistency.

I'd love to see a PR clean up all of this logic and move it to one place. And then be applied to this link as well.

Would you like to do this in this PR or open an issue outlining the refactor? 

> And I also haven't added other translations (es, pt, fr, zh). I don't speak these languages, but I can use Google Translator if you want.

Nope - skip the translations for now!

File: app/assets/images/icons/outline/calendar_days.svg
Comment (jkostolansky) [reply app/assets/images/icons/outline/calendar_days.svg:1]: Done.

File: app/components/businesses/scheduling_link_component.rb (hunk 2 above)
Comment (jkostolansky) [reply app/components/businesses/scheduling_link_component.rb]: Nice, changed that.

**Comment by jkostolansky** [comment]
### Changed

> Can you please add the NEW functionality to this field?

Totally missed that. It's there now.

> I'd love for the Core Developer to be seeded with a scheduling link.

Done.

> I'd like the design to be a button group, similar to the admin section on a developer profile. Let me know if you'd like to apply that change or if you want me to take the first pass.

There is only one button, so I just styled the link as a single button for now. Is that okay?

<details>
<summary>Screenshot</summary>

CleanShot 2023-03-25 at 00 27 22

</details>

---

### To Do

> > I also haven't added basic URL validation for this attribute. I've noticed the current website attribute doesn't have any validation, too. And I've also noticed that the fields website and mastodon don't remove the protocol prefix (http/s), so it's somewhat duplicated in the form. I think these issues should be fixed together to maintain consistency.
> 
> I'd love to see a PR clean up all of this logic and move it to one place. And then be applied to this link as well.
> Would you like to do this in this PR or open an issue outlining the refactor?

Whatever works for you. 🙂

**Review by joemasilotti — APPROVED** [review]

**Comment by joemasilotti** [comment]
Thanks for your contribution @jkostolansky! I made a few design tweaks to the form. Your bounty is on the way.

> I'd love to see a PR clean up all of this logic and move it to one place. And then be applied to this link as well.

Can you open a new issue documenting this? $25 bounty, too, of course. You are welcome to work on it or not, just let me know!

**Comment by jkostolansky** [comment]
Thanks for the final fixes and finishing touches. 👍

> Can you open a new issue documenting this? $25 bounty, too, of course. You are welcome to work on it or not, just let me know!

Sure! I will open an issue and I can work on that.

**Pull request merged**